contracts/andromeda_mission/src/testing/mod.rs

```diff
@@ -0,0 +1,585 @@
+use crate::{contract::*, state::ADO_ADDRESSES};
+use andromeda_protocol::{
+    mission::{ExecuteMsg, InstantiateMsg, MissionComponent},
+    testing::mock_querier::mock_dependencies_custom,
+};
+use common::{ado_base::AndromedaMsg, error::ContractError};
+use cosmwasm_std::{
+    testing::{mock_dependencies, mock_env, mock_info},
+    to_binary, Addr, CosmosMsg, StdError, WasmMsg,
+};
+
+#[test]
+fn test_empty_instantiation() {
+    let mut deps = mock_dependencies(&[]);
+
+    let msg = InstantiateMsg {
+        operators: vec![],
+        mission: vec![],
+        xfer_ado_ownership: false,
+        name: String::from("Some Mission"),
+        primitive_contract: String::from("primitive_contract"),
+    };
+    let info = mock_info("creator", &[]);
+
+    // we can just call .unwrap() to assert this was a success
+    let res = instantiate(deps.as_mut(), mock_env(), info, msg).unwrap();
+    assert_eq!(0, res.messages.len());
+}
+
+#[test]
+fn test_instantiation() {
+    let mut deps = mock_dependencies_custom(&[]);
+
+    let msg = InstantiateMsg {
+        operators: vec![],
+        mission: vec![MissionComponent {
+            name: "token".to_string(),
+            ado_type: "cw721".to_string(),
+            instantiate_msg: to_binary(&true).unwrap(),
+        }],
+        xfer_ado_ownership: false,
+        name: String::from("Some Mission"),
+        primitive_contract: String::from("primitive_contract"),
+    };
+    let info = mock_info("creator", &[]);
+
+    let res = instantiate(deps.as_mut(), mock_env(), info, msg).unwrap();
+    assert_eq!(1, res.messages.len());
+    match res.messages[0].msg.clone() {
+        CosmosMsg::Wasm(WasmMsg::Instantiate { code_id, msg, .. }) => {
+            assert_eq!(4, code_id);
+            assert_eq!(to_binary(&true).unwrap(), msg)
+        }
+        _ => panic!("Invalid msg type"),
+    }
+}
+
+// TODO: Figure out how to test this
```

Comment:
What is the problem you are facing with this test?

Reply:
The ADO name/address mapping is already stored and there's a duplicate check in the method used to add a component.